File: PersonajesDeMario.py
from abc import ABC, abstractmethod

class Personaje(ABC):
    def __init__(self, nombre: str, nivel: int, puntos_vida: int):
        self.nombre = nombre
        self.nivel = nivel
        self.puntos_vida = puntos_vida

    def recibir_dano(self, d: int):
        self.puntos_vida = max(0, self.puntos_vida - d)

    @abstractmethod
    def atacar(self, obj: "Personaje"):
        pass

    def __str__(self):
        return f"{self.nombre} (nivel {self.nivel}, vida {self.puntos_vida})"


class Mario(Personaje):
    def __init__(self):
        super().__init__("Mario", 3, 100)
        self.tiene_star = False

    def atacar(self, obj: Personaje):
        obj.recibir_dano(20)


class Luigi(Personaje):
    def __init__(self):
        super().__init__("Luigi", 3, 100)
        self.tiene_star = False

    def atacar(self, obj: Personaje):
        obj.recibir_dano(20)


class Toad(Personaje):
    def __init__(self):
        super().__init__("Toad", 1, 50)
        self.tiene_star = False

    def atacar(self, obj: Personaje):
        obj.recibir_dano(20)


class Orco(Personaje):
    def __init__(self):
        super().__init__("Orco", 2, 200)

    def atacar(self, obj: Personaje):
        obj.recibir_dano(15)


if __name__ == "__main__":
    equipo: list[Personaje] = [Mario(), Luigi(), Toad()]
    orco = Orco()

    print(f"Antes: {orco}")

    for p in equipo:
        p.atacar(orco)
        print(f"{p.nombre} atacó → {orco}")

File: PersonajesDeMario.java
import java.util.ArrayList;

public class PersonajesDeMario {

    public abstract static class Personaje {
        protected String nombre;
        protected int nivel, puntosVida;

        public Personaje(String nombre, int nivel, int vida) {
            this.nombre = nombre;
            this.nivel = nivel;
            this.puntosVida = vida;
        }

        public void recibirDano(int d) {
            puntosVida = Math.max(0, puntosVida - d);
        }

        public abstract void atacar(Personaje obj);

        @Override
        public String toString() {
            return nombre + " (nivel " + nivel + ", vida " + puntosVida + ")";
        }
    }

    public static class Mario extends Personaje {
        private boolean tieneStar;

        public Mario() {
            super("Mario", 3, 100);
            this.tieneStar = false;
        }

        @Override
        public void atacar(Personaje obj) {
            obj.recibirDano(20);
        }
    }

    public static class Luigi extends Personaje {
        private boolean tieneStar;

        public Luigi() {
            super("Luigi", 3, 100);
            this.tieneStar = false;
        }

        @Override
        public void atacar(Personaje obj) {
            obj.recibirDano(20);
        }
    }

    public static class Toad extends Personaje {
        private boolean tieneStar;

        public Toad() {
            super("Toad", 1, 50);
            this.tieneStar = false;
        }

        @Override
        public void atacar(Personaje obj) {
            obj.recibirDano(20);
        }
    }

    // Enemigo de ejemplo para recibir los ataques
    public static class Orco extends Personaje {
        public Orco() {
            super("Orco", 2, 200);
        }

        @Override
        public void atacar(Personaje obj) {
            obj.recibirDano(15);
        }
    }

    public static void main(String[] args) {
        ArrayList<Personaje> equipo = new ArrayList<>();
        equipo.add(new Mario());   // ← corregido
        equipo.add(new Luigi());
        equipo.add(new Toad());

        Personaje orco = new Orco(); // ← definido correctamente
        System.out.println("Antes: " + orco);

        for (Personaje p : equipo) {
            p.atacar(orco);          // ← POLIMORFISMO ✨
            System.out.println(p.nombre + " atacó → " + orco);
        }
    }
}